main/src/library/Array.flix

```diff
@@ -57,10 +57,1390 @@ namespace Array {
+    /// That is, the result is of the form: `[..., f(xn-1, f(xn, s)), f(xn, s), s]`.
+    ///
+    pub def scanRight(f: (a, b) -> b & e, s: b, a: Array[a]): Array[b] & Impure =
+        let len = length(a);
+        let b = new(s, len+1);
+        scanRightHelper(f, s, a, b, len-1);
+        b
+
+    ///
+    /// Helper for `scanRight`.
+    ///
+    def scanRightHelper(f: (a, b) -> b & e, s: b, a: Array[a], b: Array[b], i: Int32) : Unit & Impure =
+        if (i < 0)
+            ()
+        else {
+            let s1 = f(a[i], s);
+            b[i] = s1;
+            scanRightHelper(f, s1, a, b, i-1)
+        }
+
+    ///
+    /// Returns the result of applying `f` to every element in `a`.
+    ///
+    /// The result is a new array.
+    ///
+    pub def map(f: a -> b & e, a: Array[a]): Array[b] & Impure =
+        let len = length(a);
+        init(i -> f(a[i]), len)
+
+    ///
+    /// Apply `f` to every element in array `a`. Array `a` is mutated.
+    ///
+    pub def mapInPlace(f: a -> a, a: Array[a]): Unit & Impure =
+        mapInPlaceHelper(f, length(a), a, 0)
+
+    ///
+    /// Helper of `mapInPlace`
+    ///
+    def mapInPlaceHelper(f: a -> a, len: Int32, a: Array[a], i: Int32): Unit & Impure =
+        if (i >= len)
+            ()
+        else {
+            a[i] = f(a[i]);
+            mapInPlaceHelper(f, len, a, i+1)
+        }
+
+    ///
+    /// Returns the result of applying `f` to every element in `a` along with that element's index.
+    ///
+    /// That is, the result is of the form: `[ f(a[0], 0), f(a[1], 1), ... ]`.
+    ///
+    pub def mapWithIndex(f: (a, Int32) -> b & e, a: Array[a]): Array[b] & Impure =
+        let len = length(a);
+        init(i -> f(a[i], i), len)
+
+
+    ///
+    /// Apply `f` to every element in array `a` along with that element's index. Array `a` is mutated.
+    ///
+    pub def mapWithIndexInPlace(f: (a, Int32) -> a, a: Array[a]): Unit & Impure =
+        mapWithIndexInPlaceHelper(f, length(a), a, 0)
+
+    ///
+    /// Helper of `mapWithIndexInPlace`
+    ///
+    def mapWithIndexInPlaceHelper(f: (a, Int32) -> a, len: Int32, a: Array[a], i: Int32): Unit & Impure =
+        if (i >= len)
+            ()
+        else {
+            a[i] = f(a[i], i);
+            mapWithIndexInPlaceHelper(f, len, a, i+1)
+        }
+
+    ///
+    /// Returns the result of applying `f` to every element in `a` and concatenating the results.
+    ///
+    pub def flatMap(f: a -> Array[b] & e, a: Array[a]): Array[b] & Impure =
+        let len = length(a);
+        init(i -> f(a[i]), len) |> flatten
+
+    ///
+    /// Returns the reverse of `a`.
+    ///
+    pub def reverse(a: Array[a]): Array[a] & Impure=
+        let len = length(a);
+        let end = len - 1;
+        init(i -> a[end - i], len)
+        
+    ///
+    /// Returns the reverse of `a`.
+    ///
+    pub def reverseInPlace(a: Array[a]): Unit & Impure =
+        let len = length(a);
+        reverseInPlaceHelper(a, len / 2, 0, len - 1)
```

Comment:
Here it seems like `inPlace` is better, because what name can you come up with that signifies "reverse" but "in a mutating way".